pip/pip-344.md

```diff
@@ -0,0 +1,87 @@
+
+**Now `pulsarClient.getPartitionsForTopic`'s behavior**
+| case | broker allow `auto-create` | param allow <br> `create if not exists` | non-partitioned topic | partitioned topic |  current behavior |
+| --- | --- | --- | --- | --- | --- |
+| 1 | `true/false` | `true/false` | `exists: true` | | REST API: `partitions: 0`<br> Client API: `partitions: 0` |
+| 2 | `true/false` | `true/false` | | `exists: true` <br> `partitions: 3` | REST API: `partitions: 3`<br> Client API: `partitions: 3` |
+| 3 | `true` | `true` | | | REST API: <br> &nbsp;&nbsp;- `create new: true` <br> &nbsp;&nbsp;- `partitions: 3` <br> Client API: <br> &nbsp;&nbsp;- `create new: true` <br> &nbsp;&nbsp;- `partitions: 3` <br> |
+| 4 | `true` | `false` | | | REST API: <br> &nbsp;&nbsp;- `create new: false` <br> &nbsp;&nbsp;- `partitions: 0` <br> Client API: <br> &nbsp;&nbsp;not support <br> |
+| 5 | `false` | `true` | | | REST API: <br> &nbsp;&nbsp;- `create new: false` <br> &nbsp;&nbsp;- `partitions: 0` <br> Client API: <br> &nbsp;&nbsp;- `create new: false` <br> &nbsp;&nbsp;- `partitions: 0` <br> |
+
+# Motivation
+
+The param `create if not exists` of the Client API is always `true.`
+
+- For case 4 of `pulsarClient.getPartitionsForTopic`'s behavior, it always tries to create the partitioned metadata, but the API name is `getxxx`.
+- For case 5 of `pulsarClient.getPartitionsForTopic`'s behavior, it returns a `0` partitioned metadata, but the topic does not exist. For the correct behavior of this case, we had discussed [here](https://github.com/apache/pulsar/issues/8813) before.
+
+# Goals
+
+Correct the behaviors of case 4 and case 5.
+
+- Do not create the partitioned metadata when calling `pulsarClient.getPartitionsForTopic`. The partitioned metadata will only be created when consumers/producers are trying to register.
+- Instead of returning a `0` partitioned metadata, respond to a not found error when calling `pulsarClient.getPartitionsForTopic` if the topic does not exist.
+
+# Detailed Design
+
+## Public-facing Changes
+
+When you call the public API `pulsarClient.getPartitionsForTopic`, pulsar will not create the partitioned metadata anymore.
+- [flink-connector-pulsar](https://github.com/apache/flink-connector-pulsar/blob/main/flink-connector-pulsar/src/main/java/org/apache/flink/connector/pulsar/sink/writer/topic/ProducerRegister.java#L221-L227) is using this API to create partitioned topic metadata.
+
+### Public API
+**LookupService.java**
+```
+// This API existed before. Not change it, thus ensuring compatibility.
+CompletableFuture<PartitionedTopicMetadata> getPartitionedTopicMetadata(TopicName topicName);
+
++ // A new API that contains an additional param "createIfAutoCreationEnabled."
++ CompletableFuture<PartitionedTopicMetadata> getPartitionedTopicMetadata(TopicName topicName, boolean createIfAutoCreationEnabled);
```

Comment:
`LookupService` is a public interface under the `pulsar-client` module, not the `pulsar-client-api` module. We should not encourage users to leverage such APIs because the semantics might change in the next minor release.

So I suggest just changing the semantic of the original `getPartitionedTopicMetadata` rather than adding a new API and leaving the old API never used internally.

Reply:
> LookupService is a public interface under the pulsar-client module, not the pulsar-client-api module. We should not encourage users to leverage such APIs because the semantics might change in the next minor release.
> So I suggest just changing the semantic of the original getPartitionedTopicMetadata rather than adding a new API and leaving the old API never used internally.

Since the extra components/pulgins may use the API `lookupService. getPartitionedTopicMetadata`, keeping it there can guarantee the successful compilation of the extra components/plugins.

Reply:
**We should not provide such API stability guarantees.**

Guaranteeing such API could only make code ugly and harder to maintain. If we make such decisions, many PRs should be challenged. For example, https://github.com/apache/pulsar/pull/21465/, which modifies a public interface `CompactedTopic`, is even cherry-picked into release branches.

The author of 3rd party systems should be responsible to make the compilation succeed if he/she decided to use an internal API, even it's a public interface. Just like I modified the KSN code (https://github.com/streamnative/ksn/pull/212) to adapt the API change of https://github.com/apache/pulsar/pull/21465

Reply:
Sure, I have split it into two separate APIs

<img width="985" alt="Screenshot 2024-03-20 at 20 50 18" src="https://github.com/apache/pulsar/assets/25195800/19017ecc-8898-4203-a69f-0a63433ef58e">

Reply:
We can modiy the implementation for the old API as below, so it can be easier to maintain. 

Since the old API has been marked `@Deprecated`, we can remove it in the future, right?

```java
@Deprecated
default CompletableFuture<List<String>> getPartitionsForTopic(String topic) {
   // call the new API.
   getPartitionsForTopic(topic, true);
}
```

Reply:
I don't mean `getPartitionsForTopic`, I mean `LookupService#getPartitionedTopicMetadata`. Why must we retain an internal method (even if it's a public interface) that is never used?

Reply:
> I don't mean `getPartitionsForTopic`, I mean `LookupService#getPartitionedTopicMetadata`.

It is the same as `PulsarClient. getPartitionsForTopic`

```java
+ @Deprecated
- CompletableFuture<PartitionedTopicMetadata> getPartitionedTopicMetadata(TopicName topicName);
+ default CompletableFuture<List<String>> getPartitionedTopicMetadata(String topic) {
+    getPartitionsForTopic(topic, true);
+ }
```

> Why must we retain an internal method (even if it's a public interface) that is never used?

Since it is a public API, I do not know whether it was used by others or not, right? And it can be easier to maintain now, and we can remove it in the future, right?

Reply:
@BewareMyPower 

OK, I removed the old API; if someone says they are using it before the PIP is done, we add it back.

Reply:
> Since it is a public API, I do not know whether it was used by others or not, right?

Let's take back to the example I mentioned https://github.com/apache/pulsar/pull/21465. KSN depends on the `CompactedTopic#asyncReadEntriesOrWait` API, should we add the original API back? There are many other examples in KoP/KSN that I have to modify the code caused by the API changes in Apache Pulsar.

I opened a discussion here: https://lists.apache.org/thread/v4bswc3byqtvpl3h4cjpvzm8sbsy0wq9